File: src/c47/distributions/gev.c
/* This file is part of 43S.
 *
 * 43S is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 43S is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 43S.  If not, see <http://www.gnu.org/licenses/>.
 */

/********************************************//**
 * \file gev.c
 ***********************************************/

#include "distributions/gev.h"

#include "constantPointers.h"
#include "error.h"
#include "flags.h"
#include "mathematics/comparisonReals.h"
#include "mathematics/expMOne.h"
#include "mathematics/wp34s.h"
#include "registers.h"
#include "registerValueConversions.h"
#include "c47.h"

static bool_t checkParamGEV(real_t *x, real_t *mu, real_t *sigma, real_t *xi, bool_t qf) {
  real_t t;

  if (!getRegisterAsReal(REGISTER_X, x)
      || !getRegisterAsReal(REGISTER_M, mu)
      || !getRegisterAsReal(REGISTER_S, sigma)
      || !getRegisterAsReal(REGISTER_Q, xi))
    goto err;

  if(realIsNegative(x)) {
    displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
    #if(EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function checkParamGeometric:", "cannot calculate for x < 0", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    goto err;
  }
  else if(realIsZero(sigma) || realIsNegative(sigma)) {
    displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
    #if(EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function checkParamGEV:", "the parameter sigma must be positive", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    goto err;
  }

  // Check support range
  if (qf || realIsZero(xi))
    return true;                                // xi = 0, full line
  realSubtract(mu, sigma, &t, &ctxtReal39);
  realDivide(&t, xi, &t, &ctxtReal39);
  if (realIsNegative(xi))
    return realCompareLessEqual(x, &t);         // xi < 0, (-inf, (mu-sigma)/xi]
  return realCompareGreaterEqual(x, &t);        // xi > 0, [(mu-sigma)/xi, +inf)

  err:
  if(getSystemFlag(FLAG_SPCRES)) {
    convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
  }
  return false;
}

static void tGEV(const real_t *x, const real_t *mu, const real_t *sigma, const real_t *xi, real_t *t) {
  real_t z;

  realSubtract(x, mu, &z, &ctxtReal39);
  realDivide(&z, sigma, &z, &ctxtReal39);   // z = (x - mu) / sigma

  if (!realIsZero(xi)) {
    realMultiply(&z, xi, &z, &ctxtReal39);    // z = xi (x - mu) / sigma
    WP34S_Ln1P(&z, &z, &ctxtReal39);
    realDivide(&z, xi, &z, &ctxtReal39);
  }
  realChangeSign(&z);
  realExp(&z, t, &ctxtReal39);
}

void fnGEVP(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi, t, u, v;

  if (!checkParamGEV(&x, &mu, &sigma, &xi, false))
    return;
  tGEV(&x, &mu, &sigma, &xi, &t);
  realAdd(&t, const_1, &u, &ctxtReal39);
  realPower(&t, &u, &v, &ctxtReal39);       // v= t^(xi+1)
  realDivide(&v, &sigma, &u, &ctxtReal39);   // u = t^(xi+1) / sigma
  realChangeSign(&t);
  realExp(&t, &v, &ctxtReal39);             // v = e^-t
  realMultiply(&v, &u, &x, &ctxtReal39);
  convertRealToResultRegister(&x, REGISTER_X, amNone);
  adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
}

static void lowerLnGEV(const real_t *x, const real_t *mu, const real_t *sigma, const real_t *xi, real_t *z) {
  tGEV(x, mu, sigma, xi, z);
  realChangeSign(z);
}

void fnGEVL(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

  if (!checkParamGEV(&x, &mu, &sigma, &xi, false))
    return;
  lowerLnGEV(&x, &mu, &sigma, &xi, &x);
  realExp(&x, &x, &ctxtReal39);
  convertRealToResultRegister(&x, REGISTER_X, amNone);
  adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
}

void fnGEVR(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

  if (!checkParamGEV(&x, &mu, &sigma, &xi, false))
    return;
  lowerLnGEV(&x, &mu, &sigma, &xi, &x);
  realExpM1(&x, &x, &ctxtReal39);
  realChangeSign(&x);
  convertRealToResultRegister(&x, REGISTER_X, amNone);
  adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
}

void fnGEVI(uint16_t unusedButMandatoryParameter) {
  real_t p, mu, sigma, xi, t, x;

  if (!checkParamGEV(&p, &mu, &sigma, &xi, true))
    return;

  if(realCompareLessEqual(&p, const_0) || realCompareGreaterEqual(&p, const_1)) {
    displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
    #if(EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function fnExponentialI:", "the argument must be 0 < x < 1", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    if(getSystemFlag(FLAG_SPCRES)) {
      convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
    }
    return;
  }
  tGEV(&p, &mu, &sigma, &xi, &t);

  convertRealToResultRegister(&x, REGISTER_X, amNone);
  adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
}

File: zig_src/shell/distributions/gev.zig
// SPDX-License-Identifier: GPL-3.0-only
//
// Zig owner for the Generalised Extreme Value distribution commands, porting
// `src/c47/distributions/gev.c`. The four fn* entry points are dispatched from
// the items.c function table (exported via frontier.zig). All decNumber/register
// access goes through frontier_distribution_runtime. The testSuite has no GEV
// dispatch entry, so these are verified by the distributions parity harness.
//
// The port reproduces upstream behaviour exactly, including the PDF's t^(t+1)
// exponent (upstream computes realPower(t, t+1); its comment says t^(xi+1)).

const dr = @import("distribution_runtime.zig");
const real_t = dr.real_t;

// noinline: called from all four entry points; a single shared copy keeps the
// flash-limited full package (pkg1) within budget.
noinline fn checkParamGEV(x: *real_t, mu: *real_t, sigma: *real_t, xi: *real_t, qf: bool) linksection(dr.code_section) bool {
    const ctx = &dr.ctxtReal39;
    var t: real_t = undefined;

    if (!dr.saveLastX()) {
        return false;
    }

    if (!dr.getRegisterAsReal(dr.REGISTER_X, x) or !dr.getRegisterAsReal(dr.REGISTER_M, mu) or
        !dr.getRegisterAsReal(dr.REGISTER_S, sigma) or !dr.getRegisterAsReal(dr.REGISTER_Q, xi))
    {
        dr.specialResultNaN();
        return false;
    }

    if (dr.realIsNegative(x)) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamGEV:", "cannot calculate for x < 0", null, null);
        dr.specialResultNaN();
        return false;
    } else if (dr.realIsZero(sigma) or dr.realIsNegative(sigma)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamGEV:", "the parameter sigma must be positive", null, null);
        dr.specialResultNaN();
        return false;
    }

    // Check support range.
    if (qf or dr.realIsZero(xi)) {
        return true; // xi = 0, full line
    }
    dr.realSubtract(mu, sigma, &t, ctx);
    dr.realDivide(&t, xi, &t, ctx);
    if (dr.realIsNegative(xi)) {
        return dr.realCompareLessEqual(x, &t); // xi < 0, (-inf, (mu-sigma)/xi]
    }
    return dr.realCompareGreaterEqual(x, &t); // xi > 0, [(mu-sigma)/xi, +inf)
}

noinline fn tGEV(x: *const real_t, mu: *const real_t, sigma: *const real_t, xi: *const real_t, t: *real_t) linksection(dr.code_section) void {
    const ctx = &dr.ctxtReal39;
    var z: real_t = undefined;

    dr.realSubtract(x, mu, &z, ctx);
    dr.realDivide(&z, sigma, &z, ctx); // z = (x - mu) / sigma

    if (!dr.realIsZero(xi)) {
        dr.realMultiply(&z, xi, &z, ctx); // z = xi (x - mu) / sigma
        dr.WP34S_Ln1P(&z, &z, ctx);
        dr.realDivide(&z, xi, &z, ctx);
    }
    dr.realChangeSign(&z);
    dr.realExponential(&z, t, ctx);
}

fn lowerLnGEV(x: *const real_t, mu: *const real_t, sigma: *const real_t, xi: *const real_t, z: *real_t) linksection(dr.code_section) void {
    tGEV(x, mu, sigma, xi, z);
    dr.realChangeSign(z);
}

pub fn gevP(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var xi: real_t = undefined;
    var t: real_t = undefined;
    var u: real_t = undefined;
    var v: real_t = undefined;

    if (!checkParamGEV(&x, &mu, &sigma, &xi, false)) {
        return;
    }
    tGEV(&x, &mu, &sigma, &xi, &t);
    dr.realAdd(&t, dr.const1(), &u, ctx);
    dr.realPow(&t, &u, &v, ctx); // upstream: v = t^(t+1)
    dr.realDivide(&v, &sigma, &u, ctx);
    dr.realChangeSign(&t);
    dr.realExponential(&t, &v, ctx); // v = e^-t
    dr.realMultiply(&v, &u, &x, ctx);
    dr.storeResult(&x);
}

pub fn gevL(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var xi: real_t = undefined;

    if (!checkParamGEV(&x, &mu, &sigma, &xi, false)) {
        return;
    }
    lowerLnGEV(&x, &mu, &sigma, &xi, &x);
    dr.realExponential(&x, &x, ctx);
    dr.storeResult(&x);
}

pub fn gevR(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var xi: real_t = undefined;

    if (!checkParamGEV(&x, &mu, &sigma, &xi, false)) {
        return;
    }
    lowerLnGEV(&x, &mu, &sigma, &xi, &x);
    dr.realExpM1(&x, &x, ctx);
    dr.realChangeSign(&x);
    dr.storeResult(&x);
}

pub fn gevI(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var p: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var xi: real_t = undefined;
    var x: real_t = undefined;
    var lnp: real_t = undefined;

    if (!checkParamGEV(&p, &mu, &sigma, &xi, true)) {
        return;
    }

    var domain_okay = dr.realGreaterThan(&p, dr.const0()) or dr.realLessThan(&p, dr.const1());
    if (!domain_okay and !dr.realIsZero(&xi)) {
        if (dr.realIsNegative(&xi)) {
            domain_okay = dr.realEqual(&p, dr.const1());
        } else {
            domain_okay = dr.realIsZero(&p);
        }
    }

    if (!domain_okay) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function fnGEVI:", "the argument out of range", null, null);
        dr.specialResultNaN();
        return;
    }

    dr.WP34S_Ln(&p, &lnp, ctx);
    dr.realChangeSign(&lnp);
    if (dr.realIsZero(&xi)) {
        dr.WP34S_Ln(&lnp, &p, ctx);
        dr.realChangeSign(&p);
        dr.realFMA(&sigma, &p, &mu, &x, ctx);
    } else {
        dr.realMinus(&xi, &p, ctx);
        dr.powerReal(&lnp, &p, &lnp, ctx); // lnp = (-ln p)^(-xi)
        dr.realSubtract(&lnp, dr.const1(), &p, ctx);
        dr.realDivide(&p, &xi, &lnp, ctx);
        dr.realFMA(&lnp, &sigma, &mu, &x, ctx);
    }
    dr.storeResult(&x);
}
